filters/auth/grantcookie.go

```diff
@@ -0,0 +1,134 @@
+package auth
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/zalando/skipper/secrets"
+	"golang.org/x/oauth2"
+)
+
+type cookie struct {
+	AccessToken  string    `json:"access_token"`
+	RefreshToken string    `json:"refresh_token"`
+	Expiry       time.Time `json:"expiry,omitempty"`
+	RefreshAfter time.Time `json:"refresh_after,omitempty"`
+}
+
+func refreshAfter(expiry time.Time) time.Time {
+	now := time.Now()
+	d := expiry.Sub(now)
+	if d <= 0 {
+		return now
+	}
+
+	d /= 10
```

Comment:
Why is it divided by 10 here?

Reply:
@aryszka could you give some background on this one?

Reply:
if refresh only after expiry, then there is a high chance that we miss the time window with the refresh when the token is still valid. Instead, we refresh on user activity. But to avoid refreshing on every request, we only refresh when the token is expire-expiry/10 old. This results 10% precision, which is tolerable from a human workflow's perspective. Just like it is tolerable to maximize the refresh rate in 1 minute.

Example, refreshed:

1. I login and get a token at 9:15 that expires at 9:45 (30 minutes expiration time)
2. I make some authenticated requests to a page between 9:15 and 9:18, no refresh
3. I make a request at 9:19, my token is still valid, and I get a refreshed token, and I will not need to login until 9:49, instead of 9:45
4. I can continue my workflow with step 2

Example, expired:

1. I login and get a token at 9:15 that expires at 9:45 (30 minutes expiration time)
2. I make some requests to a page between 9:15 and 9:17, but then I finish
3. I go back to the page at 9:46. My token is expired, and I have to login again
4. I understand this, because I didn't make any requests for 29 minutes, which I perceive as approximately the expiration time of 30 minutes
5. I can continue my workflow, but only from step 1

Does this make sense?

I think that deserves a code comment, though, I agree with that.

Reply:
I checked: https://tools.ietf.org/html/rfc6749#section-1.5

I see that this approach is not required, and the refresh token can be used regardless of the validity of the access token. This means that we can just use the expiration of the token as is, and there is no need for the above flow.

Reply:
As discussed with @aryszka we can just drop this func and its use

Reply:
Dropped in a5d702c